ssl/s23_clnt.c

```diff
@@ -375,12 +357,13 @@ static int ssl23_client_hello(SSL *s)
 
     buf = (unsigned char *)s->init_buf->data;
     if (s->state == SSL23_ST_CW_CLNT_HELLO_A) {
-        /*
-         * Since we're sending s23 client hello, we're not reusing a session, as
-         * we'd be using the method from the saved session instead
-         */
-        if (!ssl_get_new_session(s, 0)) {
-            return -1;
+      /* Check if the session is resumable. If not, drop it. */
+        if (s->session == NULL ||
+            s->session->ssl_version > version ||
+            s->session->session_id_length == 0 ||
+            s->session->not_resumable) {
+            if (!ssl_get_new_session(s, 0))
```

Comment:
Thinking more about it, I won't be able to land second patch, because it will cause ABI incompatibility.